compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -545,67 +545,109 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
-            *tgt = Ok(v);
+            let imm = transmute_immediate(bx, src, from_scalar, from_bty, to_scalar, to_bty);
+            *tgt = Ok(imm);
         };
 
         match (operand.val, operand.layout.backend_repr) {
-            (OperandValue::ZeroSized, _) => {
-                debug_assert_eq!(field_layout.size, Size::ZERO);
-            }
+            (OperandValue::ZeroSized, _) if expect_zst => {}
             (OperandValue::Immediate(v), BackendRepr::Scalar(from_scalar)) => match &mut self.val {
-                OperandValue::Immediate(val @ Err(_)) => {
-                    debug_assert_eq!(field_offset, Size::ZERO);
+                OperandValue::Immediate(val @ Err(_)) if is_zero_offset => {
                     update(val, v, from_scalar);
-                    //*val = Ok(v);
                 }
-                OperandValue::Pair(fst @ Err(_), _) if field_offset == Size::ZERO => {
+                OperandValue::Pair(fst @ Err(_), _) if is_zero_offset => {
                     update(fst, v, from_scalar);
-                    //*fst = Ok(v);
                 }
-                OperandValue::Pair(_, snd @ Err(_)) if field_offset != Size::ZERO => {
+                OperandValue::Pair(_, snd @ Err(_)) if !is_zero_offset => {
                     update(snd, v, from_scalar);
-                    //*snd = Ok(v);
                 }
-                _ => bug!("Tried to insert {operand:?} into field {f:?} of {self:?}"),
+                _ => bug!("Tried to insert {operand:?} into {v:?}.{f:?} of {self:?}"),
             },
             (OperandValue::Pair(a, b), BackendRepr::ScalarPair(from_sa, from_sb)) => {
                 match &mut self.val {
                     OperandValue::Pair(fst @ Err(_), snd @ Err(_)) => {
                         update(fst, a, from_sa);
-                        //*fst = Ok(a);
                         update(snd, b, from_sb);
-                        //*snd = Ok(b);
                     }
-                    _ => bug!("Tried to insert {operand:?} into field {f:?} of {self:?}"),
+                    _ => bug!("Tried to insert {operand:?} into {v:?}.{f:?} of {self:?}"),
                 }
             }
-            _ => bug!("Unsupported operand {operand:?} inserting into field {f:?} of {self:?}"),
+            _ => bug!("Unsupported operand {operand:?} inserting into {v:?}.{f:?} of {self:?}"),
+        }
+    }
+
+    pub(super) fn insert_imm(&mut self, f: FieldIdx, imm: V) {
+        let field_offset = self.layout.fields.offset(f.as_usize());
+        let is_zero_offset = field_offset == Size::ZERO;
+        match &mut self.val {
+            OperandValue::Immediate(val @ Err(_)) if is_zero_offset => {
+                *val = Ok(imm);
+            }
+            OperandValue::Pair(fst @ Err(_), _) if is_zero_offset => {
+                *fst = Ok(imm);
+            }
+            OperandValue::Pair(_, snd @ Err(_)) if !is_zero_offset => {
```

Comment:
intrusive thoughts about whether these are actually required to be ordered in this way